(window.webpackJsonp=window.webpackJsonp||[]).push([[471],{1277:function(r,e,t){"use strict";t.r(e);var n=t(28),o=Object(n.a)({},(function(){var r=this,e=r.$createElement,t=r._self._c||e;return t("ContentSlotsDistributor",{attrs:{"slot-key":r.$parent.slotKey}},[t("h2",{attrs:{id:"代码执行的错误捕获"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#代码执行的错误捕获"}},[r._v("#")]),r._v(" 代码执行的错误捕获")]),r._v(" "),t("p",[r._v("try…catch")]),r._v(" "),t("ul",[t("li",[r._v("能捕获到代码执行的错误")]),r._v(" "),t("li",[r._v("捕获不到语法的错误")]),r._v(" "),t("li",[r._v("无法处理异步中的错误")]),r._v(" "),t("li",[r._v("使用try…catch包裹，影响代码可读性")])]),r._v(" "),t("p",[r._v("window.onerror")]),r._v(" "),t("ul",[t("li",[r._v("无论是异步还是非异步错误，onerror都能捕获到运行时错误")]),r._v(" "),t("li",[r._v("onerror主要是来捕获预料之外的错误，而try...catch则是用来在可预见情况下监控特定的错误，两者结合使用更加高效")]),r._v(" "),t("li",[r._v("window.onerror函数只有在返回true的时候，异常才不会向上抛出，否则即使知道异常的发生控制台还是会显示Uncaugth Error:xxx")]),r._v(" "),t("li",[r._v("当遇到"),t("code",[r._v('<img src="./404.png">')]),r._v("报404网络请求异常的时候，onerror是无法帮助我们捕获到异常的")])]),r._v(" "),t("p",[r._v("缺点：监听不到资源加载的报错onerror，事件处理函数只能声明一次，不会重复执行多个回调")]),r._v(" "),t("p",[r._v("3.window.addEventListener('error') 可以鉴定到资源加载报错，也可以注册多个事件处理函数。window.addEventListener('error',(msg,url,row,col,error)=>{},true),但是这种方式虽然可以捕获到网络请求的异常，却无法判断Http状态是404还是其他比如500等等，所以还需要配合服务端日志才进行排查分析才可以")]),r._v(" "),t("p",[r._v("4.window.addEventListener('unhandledrejection') 捕获Promise错误，当Promise被reject且没有reject处理器的时候，会出发unhandledrejection事件；这可能发生在window下，但也可能发生在Worker中。这对于调试回退错误处理非常有用。")]),r._v(" "),t("h2",{attrs:{id:"资源加载的错误捕获"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#资源加载的错误捕获"}},[r._v("#")]),r._v(" 资源加载的错误捕获")]),r._v(" "),t("p",[r._v("1.imgObj.onerror")]),r._v(" "),t("p",[r._v("2.performance.getEntries（），捕获到成功加载的资源，对比可以间接的捕获错误")]),r._v(" "),t("p",[r._v("3.window.addEventListener('error',fn,true),会捕获但是不冒泡，所以window.onerror不会触发，捕获阶段可以触发")]),r._v(" "),t("h2",{attrs:{id:"vue-react中"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#vue-react中"}},[r._v("#")]),r._v(" vue/react中")]),r._v(" "),t("p",[r._v("vue有errorHandler，react有componentDidCatch进行错误捕获")])])}),[],!1,null,null,null);e.default=o.exports}}]);